### PR TITLE
Add support for ed25519 in IonKey (1/2)

### PR DESCRIPTION
**Pull request by gjgd**
Part 1 of adding support for Ed25519 in IonSDK:
- Add JwkEd25519 model
- Add Ed25519 test vectors
- Add methods for generating Ed25519 keys in IonKey 

Part 2 will be updating IonRequest and IonDid

### CODE REVIEW

File: tests/IonKey.spec.ts
Comment (kdenhartog) [inline tests/IonKey.spec.ts]: Why is this being typecast here? I'm thinking this workaround should probably take into account that this additional abstraction should be extended into the type system as well (e.g. change to generic Jwk type instead)

File: package.json
Comment (kdenhartog) [inline package.json]: Is it possible to anchor these against a stable branch so that the `^` doesn't affect the input.

File: lib/models/JwkEd25519.ts
Comment (kdenhartog) [inline lib/models/JwkEd25519.ts:4]: I'd think it would be safer to type the private key as a separate type so that an API can confirm that only public data is being passed in here. By adding the `d` as optional rather than as a separate interface and then also building a type union of only the public key interfaces we would be able to leverage the type system to filter private keys which is not possible with this approach.

File: lib/IonKey.ts
Comment (kdenhartog) [inline lib/IonKey.ts]:
```suggestion
    return key.crv === 'secp256k1' && key.kty === 'EC';
```

File: lib/IonKey.ts
Comment (kdenhartog) [inline lib/IonKey.ts]:
```suggestion
    return key.crv === 'Ed25519' && key.kty === 'OKP';
```

File: lib/IonKey.ts
Comment (kdenhartog) [inline lib/IonKey.ts]: I'd suggest making this of type `JsonWebKey2020` rather than `Ed25519VerificationKey2018` since the `Ed25519VerificationKey2018` suite would drop the `publicKeyJwk` in a json-ld processor.

File: package.json
Comment (csuwildcat) [reply package.json]: Would love to pull in better versions of dependencies, just need to tweak the parts of the code that touch the interfaces exposed by those versions.

Comment (gjgd) [reply package.json]: There are [no recent stable releases](https://www.npmjs.com/package/@transmute/did-key-ed25519), so I pinned these dependencies for now to avoid drift: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/ae0afb8165fc3319448ce293181ae472d0d37085

@OR13 What is the latest version that is safe to use?

File: lib/models/JwkEd25519.ts
Comment (gjgd) [reply lib/models/JwkEd25519.ts:4]: I copied the [JwkEs256K model](https://github.com/decentralized-identity/ion-sdk/blob/61750e8/lib/models/JwkEs256k.ts) which follows the same pattern. I agree with you, but this is out of scope for this change

File: lib/IonKey.ts
Comment (gjgd) [reply lib/IonKey.ts]: Ty! https://github.com/decentralized-identity/ion-sdk/pull/20/commits/cf73d00fc9604dd33a8e390a1a35b045d62b1070

File: tests/IonKey.spec.ts
Comment (gjgd) [reply tests/IonKey.spec.ts]: Added more specific type for publicKeyJwk: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/045c4254604d808fe0bc8e0cb8d0a3f2f96c8af6

File: lib/models/JwkEd25519.ts
Comment (kdenhartog) [reply lib/models/JwkEd25519.ts:4]: Sounds good I opened #22 to track this in a separate PR

**Review by kdenhartog — APPROVED** [review]

File: package.json
Comment (OR13) [inline package.json]: beware that these versions are getting kinda old... also, i think technically only the key pair class is required, not the did key wrapper around it... possible can just use  https://github.com/transmute-industries/verifiable-data/blob/main/packages/json-web-signature/src/JsonWebKey.ts

**Review by OR13 — APPROVED** [review]

File: package.json
Comment (OR13) [reply package.json]: Definitely a good idea to pin them.

current version of did key is: https://github.com/transmute-industries/did-key.js/blob/main/packages/did-key-ed25519/package.json#L5

but you can see the latest version for the key pair class here:
https://github.com/transmute-industries/verifiable-data/blob/main/packages/ed25519-key-pair/package.json#L5

The goal should always be less dependencies... IMO, it would be better for ION SDK to use the key pair classes instead of the did key wrappers... even though the wrappers are tiny... for the most part the interfaces they expose are the same...

File: package.json
Comment (gjgd) [reply package.json]: Used `@transmute/ed25519-key-pair` and `@transmute/secp256k1-key-pair` in `IonKey` class instead of their did-key counterpart: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/ac8c23664031007274c68508f848a1be785b8724

However there was one other instance of `@transmute/did-key-secp256k1` that is left to substitute: https://github.com/decentralized-identity/ion-sdk/blob/main/lib/LocalSigner.ts#L23

The `ES256K` class is not present anymore in the newest versions of the secp256k1 key libraries I could find. @OR13 Do you have a suggestion for where I can find a newer version of this class for signing?

Comment (OR13) [reply package.json]: yes :) 

https://github.com/transmute-industries/verifiable-data/blob/main/packages/jose-ld/src/JWS/jose.test.ts

https://github.com/transmute-industries/verifiable-data/blob/main/packages/jose-ld/src/JWS/__fixtures__/transmute-helper.ts

^ the reason those JOSE modules were removed is that they all did the same thing, so they were refactored to look like this... and ensure that the same code is used for each JWS, regardless of the raw signature underneath.

Comment (gjgd) [reply package.json]: Thanks! These examples were very helpful. Done: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/661da321cf76dc38f835493e536adda24661cfa1

**Comment by csuwildcat** [comment]
So if this addition makes it possible to use Ed25519 for operation keys, will there be a PR in Sidetree Core to allow them to be used in the actual nodes?

**Comment by gjgd** [comment]
My plan was to use this change to generate Ed25519 sidetree fixtures, but it can also be used to allow Ed25519 in Sidetree core node.

**Comment by thehenrytsai** [comment]
Will take a look in the next couple of days, adding @xinaxu in case he has bandwidth first.

File: lib/LocalSigner.ts
Comment (thehenrytsai) [inline lib/LocalSigner.ts]: Assigning to `privateKey` to `publicKeyJwk` here seems not right to do. At least add a comment to say that this is deliberate and has no adverse effect? Which leads to a bigger question, what's wrong with the original one liner? Is it that `@transmute/did-key-secp256k1` is being deprecated?

File: lib/models/IonPublicKeyModel.ts
Comment (thehenrytsai) [inline lib/models/IonPublicKeyModel.ts]: Not sure the strong type enforcement is right here. It is intentionally `object` as we do not want to dictate what users put in here as their public keys, they can chose whatever algorithm they want.

File: tests/vectors/inputs/jwkEd25519Private1.json
Comment (thehenrytsai) [inline tests/vectors/inputs/jwkEd25519Private1.json]: Comment for overall test input file naming:

Be consistent with naming with the other secp2561k files. ie. Have the number (1, 2 etc) in the middle as opposed to the end. I did that so that public and private key files are next to each other making it obvious that the files come in pairs.

**Comment by thehenrytsai** [comment]
@gjgd, thanks for taking the time to start to adding ED25519 support. The PR looks great overall, given that it really isn't hooked up with any thing really, so I see this more as a preparation to more PRs to come, so just a few comments, please do keep the code coverage at 100%.

File: lib/models/IonPublicKeyModel.ts
Comment (kdenhartog) [reply lib/models/IonPublicKeyModel.ts]: Within reason right? We still want to adhere to the JWK data model IIRC

Comment (thehenrytsai) [reply lib/models/IonPublicKeyModel.ts]: @kdenhartog, yes the intention is to allow any JWK, but since JWK is such an open-ended spec, with only `kty` being the required property, I ended up not creating yet another model for it in Sidetree core reference implementation, thus as well as the ION SDK. A case can be made to enforce `kty`, and you can file a separate issue if you think that's the case, but the point here is that it should not be restricted to just the above two key types.

Comment (kdenhartog) [reply lib/models/IonPublicKeyModel.ts]: Yeah makes sense - My take was similar to yours that we don't want to overly restrict it as well

**Comment by OR13** [comment]
We are blocked by this, and need a timeline for merge. @thehenrytsai are you able to approve and merge? or should we fork the implementation and push support for this here, removing the ION SDK dep?:

https://github.com/transmute-industries/sidetree.js/blob/main/packages/wallet/src/operations/update.ts#L4

We are trying to wrap this SDK to maintain interop and support ION at the same time, but we are required to support EdDSA.

**Comment by thehenrytsai** [comment]
@OR13, I reviewed the PR couple of days ago with a few comments, expecting them to be addressed or responded to before I can approve the merge.

**Comment by OR13** [comment]
@thehenrytsai thanks, I was surprised not to see a "change request" from you on the PR... you left comments, but did not request changes.

cc @gjgd do you understand the "comments that are requesting changes"?

**Comment by gjgd** [comment]
I didn't get a chance to review the comments yet, will try to do that today and make the appropriate changes

File: lib/models/IonPublicKeyModel.ts
Comment (gjgd) [reply lib/models/IonPublicKeyModel.ts]: Reverted back to `object` type: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/2224001d45926a7bd34c2f76fc6a62bab86a476e

File: tests/vectors/inputs/jwkEd25519Private1.json
Comment (gjgd) [reply tests/vectors/inputs/jwkEd25519Private1.json]: Done here: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/51a810c6d4ad4031f308b9aaaccd67de17f865f5

For context, I put it at the end because I found that the names `Ed255191`, `Ed255192`, `Ed255193` looked a bit weird (hard to distinguish the name of the algo vs the 1 2 3)

File: lib/LocalSigner.ts
Comment (gjgd) [reply lib/LocalSigner.ts]: https://github.com/decentralized-identity/ion-sdk/pull/20/commits/d63d081902df1691bc1f9b1c27f4d204e43d7519

Yes this lib was old, see https://github.com/decentralized-identity/ion-sdk/pull/20#discussion_r758438788

**Comment by gjgd** [comment]
@thehenrytsai  Thanks for the review, I believe I addressed all of the comments. Part 2 coming next

**Comment by csuwildcat** [comment]
@thehenrytsai what's the status on this? Let's sync as a WG after the holidays to ensure this PR doesn't fall to the wayside.

File: tests/vectors/inputs/jwkEd25519Private1.json
Comment (thehenrytsai) [reply tests/vectors/inputs/jwkEd25519Private1.json]: @gjgd, valid point. Maybe flipping the name to `ed255191jwk1private` would be the best compromise. But won't hold the PR on this, singing off.

**Review by thehenrytsai — APPROVED** [review]
Read your comment on file naming and it's a very good point. Probably changing the name to something like `ed25519jwk1private` across the board is the way to go, but won't hold the PR on this change.

**Pull request merged**
